### PR TITLE
Integrated Overflow chip on MST 

### PR DESCRIPTION
**Pull request by sifnoc**
There are two practical configurations, which are tested, for the overflow chip:

1. 12 bits with 21 Columns
2. 8 bits with 31 Columns

All test cases within the Merkle Sum Tree have been modified to accommodate the second overflow chip configuration.

### CODE REVIEW

File: src/chips/merkle_sum_tree.rs
Comment (enricobottazzi) [inline src/chips/merkle_sum_tree.rs:310]: I think that there's an issue here. There's no constraint in place that forces `update_value` to be equal to `left_balance` or to `right_balance`. I think this can be resolved by modifing assign function inside the overflow_chip to take `update_value: &AssignedCell<Fp, Fp>` as input and copying the assigned cell to `self.config.a`. This is a similar operation to the one performed in the `enforce_less_than()` function.

File: src/chips/overflow/mod.rs
Comment (sifnoc) [reply src/chips/overflow/mod.rs]: Fixed at [d00ed53](https://github.com/summa-dev/summa-solvency/pull/33/commits/d00ed534c4f959a33662ea6a41664e2b764599d5)

File: src/chips/merkle_sum_tree.rs
Comment (sifnoc) [reply src/chips/merkle_sum_tree.rs]: Yes, good point! we do not need to check overflow at `computed_sum`.
Updated at [d00ed53](https://github.com/summa-dev/summa-solvency/pull/33/commits/d00ed534c4f959a33662ea6a41664e2b764599d5)

File: src/chips/overflow/tests.rs
Comment (sifnoc) [reply src/chips/overflow/tests.rs]: Fixed at [d00ed53](https://github.com/summa-dev/summa-solvency/pull/33/commits/d00ed534c4f959a33662ea6a41664e2b764599d5)

File: src/chips/merkle_sum_tree.rs
Comment (sifnoc) [reply src/chips/merkle_sum_tree.rs:104]: Fixed at [b19d029](https://github.com/summa-dev/summa-solvency/pull/33/commits/b19d029acac584127dfb34474b1ba918ad2e174c) also update at [d00ed53](https://github.com/summa-dev/summa-solvency/pull/33/commits/d00ed534c4f959a33662ea6a41664e2b764599d5)

File: src/merkle_sum_tree/utils/csv_parser.rs
Comment (enricobottazzi) [inline src/merkle_sum_tree/utils/csv_parser.rs:29]: Why not 252? I think we can get there using 84*3.

File: src/chips/merkle_sum_tree.rs
Comment (sifnoc) [reply src/chips/merkle_sum_tree.rs:8]: I agree that minimizing the number of columns is beneficial, but increasing the number of bits to reduce columns in turn requires more rows for the fixed column. This implies a longer time for generating proofs. For instance, setting 20 Bits and 10 Columns resulted in a 9141s test time for `test_valid_merkle_sum_tree`. 
Therefore, practically speaking, setting the `MAXBITS` to less than 18 appears to be the maximum choice for now.

Comment (enricobottazzi) [reply src/chips/merkle_sum_tree.rs:8]: Oh damn, I see it now! 

Ok let's keep it like this for now then

File: src/merkle_sum_tree/utils/csv_parser.rs
Comment (sifnoc) [reply src/merkle_sum_tree/utils/csv_parser.rs:29]: Currently, we're using a single range table where `1 << 8` is applied to the decomposed columns for range checking. This means the overflow chip checks if the maximum number is (1 << 251) - 1. The following table might be useful for understanding:

|   	| a_0                                                            	| a_1  	| a_2  	| ... 	| a_30 	|
|---	|----------------------------------------------------------------	|------	|------	|-----	|------	|
| x 	| 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff 	| 0xff 	| 0xff 	| ... 	| 0xff 	|

In an ideal scenario, we could set the maximum value to half the modulus to avoid overflow, irrespective of the decomposed value. However, to achieve this, we'd need to create tables for range checking for each decomposed column.

|   	| a_0                                                                	| a_1  	| a_2  	| ... 	| a_32 	|
|---	|--------------------------------------------------------------------	|------	|------	|-----	|------	|
| x 	| 0x183227397098d014dc2822db40c0ac2e9419f4243cdcb848a1f0fac9f8000000 	| 0x18 	| 0x32 	| ... 	| 0x00 	|

However, I have my doubts about this approach. Each column from `a_1` to `a_30` requires a different range table. a_1 uses 0..24(0x18), a_2 uses 0..50(0x32), and so on.

One alternative that seems plausible is to use `0x1832` as the range for the table. While we could go for a larger number, we would need to set the maximum value to 2^n to fill all 252 bits. Additionally, as mentioned earlier, it seems challenging to exceed `0x183227`, which is between `1 << 20` and `1 << 21`.

The following table demonstrates how the maximum value can be checked for overflow using the 0x1832 range table:

|   	| a_0                                                                	| a_1    	| a_2    	| ... 	| a_17   	|
|---	|--------------------------------------------------------------------	|--------	|--------	|-----	|--------	|
| x 	| 0x1832183218321832183218321832183218321832183218321832183218321832 	| 0x1832 	| 0x1832 	| ... 	| 0x1832 	|

What are your thoughts on turning this idea into a new task?

File: src/chips/merkle_sum_tree.rs
Comment (enricobottazzi) [reply src/chips/merkle_sum_tree.rs:8]: @sifnoc I found a good spot setting 

```
const ACC_COLS: usize = 20;
const MAX_BITS: u8 = 10;
```

Set k to 13 and run 

```
cargo  test --release test_valid_merkle_sum_tree_with_full_prover  -- --nocapture
Start:   Creating proof
End:     Creating proof ............................................................2.137s
```

I still think that it is ok to keep it as it for now and later try tweaking for a good balance between proving time and verification cost. Just sharing it for your interest

File: src/chips/overflow/overflow_check.rs
Comment (enricobottazzi) [inline src/chips/overflow/overflow_check.rs:62]: Thanks for the refactoring. I think there's a typo here in the comment as I guess you meant 

```
a_1 * 2^16 + a_2 * 2^8 + a_3* 1.
```

Also, just to be sure, the decomposition in performed little endian, namely the most significant bits is in the first advice column?

File: src/merkle_sum_tree/utils/csv_parser.rs
Comment (enricobottazzi) [reply src/merkle_sum_tree/utils/csv_parser.rs:29]: The idea of setting this check 

```
    // throw error if balance is larger than the modulus
    if balance_acc >= BigInt::parse_bytes(MODULUS_STR.as_bytes(), 16).unwrap() {
        return Err("Balance is larger than the modulus".into());
    }
```

was to warn the CEX that if they are building a MST which balance sum is larger than the modulus, it will end up in an overflow during proof generation.

Now I see that it would be more appropriate to check that the balance should not be larger than `MAX_VALUE_STR` which is the actual range check that we perform inside the circuit for the balance. 

But the value `MAX_VALUE_STR` depends on the value assigned to `ACC_COLS` and `MAX_BITS` which are generic to the circuit and might vary. Therefore I think it makes sense to keep this check against `MODULUS_STR`.  

In the case that `MODULUS < balance_acc < MAX_VALUE` the cex would be able to build the merkle sum tree and incur in an error during proof generation, but I think this is fine.


> In an ideal scenario, we could set the maximum value to half the modulus to avoid overflow, irrespective of the decomposed value

I didn't understand this part

File: src/chips/merkle_sum_tree.rs
Comment (sifnoc) [reply src/chips/merkle_sum_tree.rs:310]: Ah, indeed! I completely overlooked this section! Thanks for bringing it to my attention
 
Fixed at [4a42627](https://github.com/summa-dev/summa-solvency/pull/33/commits/4a42627ac6bb8b1d628dc5ca9ac2deedc4d862ac)

File: src/merkle_sum_tree/utils/csv_parser.rs
Comment (sifnoc) [reply src/merkle_sum_tree/utils/csv_parser.rs:29]: > But the value `MAX_VALUE_STR` depends on the value assigned to `ACC_COLS` and `MAX_BITS` which are generic to the circuit and might vary. Therefore I think it makes sense to keep this check against `MODULUS_STR`.
> 
> In the case that `MODULUS < balance_acc < MAX_VALUE` the cex would be able to build the merkle sum tree and incur in an error during proof generation, but I think this is fine.

I agree with that, it doesn't seem like the right time to settle this. Let's revert to this.

> > In an ideal scenario, we could set the maximum value to half the modulus to avoid overflow, irrespective of the decomposed value
> 
> I didn't understand this part

If we allow balances to go beyond half the modulus, meaning the overflow chip range check exceeds half the modulus, we could run into a worst-case scenario like the one below (considering a modulus of 255 for simplicity):

Let's walk through the 'merkle_prove_layer' in the circuit with some numbers:
https://github.com/summa-dev/summa-solvency/blob/b0fa4163474ac8ee1988342e070221f5586fb7a9/src/circuits/merkle_sum_tree.rs#L58-L81

leaf_balance = 28
path_element_balance[0] = 100, next_sum = 128
path_element_balance[1] = 128, next_sum = 1

In such a situation, the final next_sum value becomes 1, which wouldn't prevent an overflow situation.

This is the rationale behind setting the ideal maximum number for the overflow chip at half the modulus

File: src/chips/overflow/overflow_check.rs
Comment (sifnoc) [reply src/chips/overflow/overflow_check.rs]: Thanks for checking it!

> Also, just to be sure, the decomposition in performed little endian, namely the most significant bits is in the first advice column?

Yes, `a_1` is MSB in the description. also, fixed at [477c0cd](https://github.com/summa-dev/summa-solvency/pull/33/commits/477c0cdc7e81bef6d3d77c89a12a59bc2307a398)

File: src/merkle_sum_tree/utils/csv_parser.rs
Comment (sifnoc) [reply src/merkle_sum_tree/utils/csv_parser.rs]: Reverted at [141d71c](https://github.com/summa-dev/summa-solvency/pull/33/commits/141d71c376f760cfccee77d0a21a6260f77f1866)

**Review by enricobottazzi — APPROVED** [review]

**Comment by sifnoc** [comment]
I plan to merge this after #29 is merged

**Review by alxkzmn — APPROVED** [review]

**Pull request merged**
